include/gmapping/gridfastslam/gridslamprocessor.hxx

```diff
@@ -25,18 +30,29 @@ inline void GridSlamProcessor::scanMatch(const double* plainReading){
 	}
     }
 
-    m_matcher.likelihoodAndScore(s, l, it->map, it->pose, plainReading);
+    m_matcher.likelihoodAndScore(s, l, m_particles[i].map, m_particles[i].pose, plainReading);
     sumScore+=score;
-    it->weight+=l;
-    it->weightSum+=l;
+    m_particles[i].weight+=l;
+    m_particles[i].weightSum+=l;
 
     //set up the selective copy of the active area
     //by detaching the areas that will be updated
     m_matcher.invalidateActiveArea();
```

Comment:
The problem with these two lines:


```m_matcher.invalidateActiveArea();
m_matcher.computeActiveArea(m_particles[i].map, m_particles[i].pose, plainReading)
```



They change map and matcher and they are designed to work sequentially. I suppose we should split this loop in two, 
something like


```//parallel
#pragma omp parallel for reduction(+:sumScore)
  for (int i = 0; i < m_particles.size(); i++){
    OrientedPoint corrected;
    double score, l, s;

    score=m_matcher.optimize(corrected, m_particles[i].map, m_particles[i].pose, plainReading);
    //    it->pose=corrected;
    if (score>m_minimumScore){
      m_particles[i].pose=corrected;
      // m_infoStream << "Pose was corrected!" << std::endl;
    } else {
	if (m_infoStream){
	  m_infoStream << "Scan Matching Failed, using odometry. Likelihood=" << l <<std::endl;
	  m_infoStream << "lp:" << m_lastPartPose.x << " "  << m_lastPartPose.y << " "<< m_lastPartPose.theta <<std::endl;
	  m_infoStream << "op:" << m_odoPose.x << " " << m_odoPose.y << " "<< m_odoPose.theta <<std::endl;
	}
    }

    m_matcher.likelihoodAndScore(s, l, m_particles[i].map, m_particles[i].pose, plainReading);
    sumScore+=score;
    m_particles[i].weight+=l;
    m_particles[i].weightSum+=l;


    m_matcher.invalidateActiveArea();
    m_matcher.computeActiveArea(m_particles[i].map, m_particles[i].pose, plainReading);
  }

//sequential
  for (int i = 0; i < m_particles.size(); i++){
    //set up the selective copy of the active area
    //by detaching the areas that will be updated
    m_matcher.invalidateActiveArea();
    m_matcher.computeActiveArea(m_particles[i].map, m_particles[i].pose, plainReading);
  }
```

Also, `m_infoStream << ...` will be garbled if two or more threads don't get enough score. I suggest either `omp critical` or plain removal of these debug lines.